src/util/amount.rs

```diff
@@ -0,0 +1,362 @@
+    }
+}
+
+impl fmt::Display for Amount {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.fmt_denomination::<Bitcoin>(f)
+    }
+}
+
+impl<P: Precision> ops::Add for Amount<P> {
+    type Output = Amount<P>;
+
+    fn add(self, rhs: Amount<P>) -> Self::Output {
+        Amount::from_inner(self.0 + rhs.0)
+    }
+}
+
+impl<P: Precision> ops::Div for Amount<P> {
+    type Output = Amount<P>;
+
+    fn div(self, rhs: Amount<P>) -> Self::Output {
+        Amount::from_inner(self.0 / rhs.0)
+    }
+}
+
+impl<P: Precision> ops::Mul for Amount<P> {
+    type Output = Amount<P>;
+
+    fn mul(self, rhs: Amount<P>) -> Self::Output {
+        Amount::from_inner(self.0 * rhs.0)
+    }
+}
+
+impl<P: Precision> ops::Sub for Amount<P> {
+    type Output = Amount<P>;
+
+    fn sub(self, rhs: Amount<P>) -> Self::Output {
+        Amount::from_inner(self.0 - rhs.0)
+    }
+}
+
+impl PartialEq for Amount {
+    fn eq(&self, other: &Amount) -> bool {
+        PartialEq::eq(&self.0, &other.0)
+    }
+}
+impl Eq for Amount {}
+
+impl PartialOrd for Amount {
+    fn partial_cmp(&self, other: &Amount) -> Option<::std::cmp::Ordering> {
+        PartialOrd::partial_cmp(&self.0, &other.0)
+    }
+}
+
+impl Ord for Amount {
+    fn cmp(&self, other: &Amount) -> ::std::cmp::Ordering {
+        Ord::cmp(&self.0, &other.0)
+    }
+}
+
+/// An error during `Amount` parsing.
+#[derive(Debug)]
+pub struct ParseAmountError(std::num::ParseFloatError);
+
+impl fmt::Display for ParseAmountError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "parse amount error: {}", self.0)
+    }
+}
+
+impl error::Error for ParseAmountError {
+    fn cause(&self) -> Option<&error::Error> {
+        Some(&self.0)
+    }
+
+    fn description(&self) -> &'static str {
+        "parse amount error"
+    }
+}
+
+impl FromStr for Amount {
+    type Err = ParseAmountError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let btc = f64::from_str(s).map_err(ParseAmountError)?;
+        Ok(btc.into_btc())
+        //Ok(IntoAmount::<Bitcoin, Satoshi>::into_amount(btc))
+    }
+}
+
+#[cfg(feature = "serde")]
+impl<'de> serde::Deserialize<'de> for Amount {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::de::Deserializer<'de>,
+    {
+        Ok(Amount::from_inner(Inner::deserialize(deserializer)?))
+    }
+}
+
+#[cfg(feature = "serde")]
+impl serde::Serialize for Amount {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::ser::Serializer,
+    {
+        Inner::serialize(&self.0, serializer)
+    }
+}
+
+//TODO(stevenroose) this trait in itself is currently of limited use since the first generic
+//parameter is always unknown, so always the awkward notation is to be used
+pub trait IntoAmount<F: Precision, T: Precision> {
+    fn into_amount(self) -> Amount<T>;
+}
+
+impl<F: Precision, T: Precision> IntoAmount<F, T> for f64 {
```

Comment:
Floating point and financial applications generally don't mix well. I'd prefer explicit `from_atto_sat(u128)` and `from_sat(u64)` functions.

Reply:
But I see you have `from_precision` for that.

Reply:
Floating points is mostly to support JSON. I think it's up to the user to know that converting a float into an amount is dangerous. Another option would be to implement `FromStr` more precisely and use it in the JSON deserializer.

Reply:
I think that would be the preferable strategy, but if we have to keep the function we should at least warn about the dangers.